### PR TITLE
Express Server

### PR DESCRIPTION
**Pull request by amorino**
Express Server with Webpack-Hot-Middleware and Dev-Middleware

### CODE REVIEW

**Comment by mxstbr** [comment]
Amazing work @amorino!

After the first glance I love the new logging!

We now see new builds and errors in the terminal

```Shell
webpack built b38b8acaa89cfa923f49 in 6980ms
webpack building...
webpack built 97f498721ca27fe8bf96 in 717ms
webpack building...
webpack built f91a819d0520f444053a in 691ms
```

and in the browser console we see the names of the modules that were hot reloaded!

```Shell
[React Transform HMR] Patching HomePage
[HMR] Updated modules:
[HMR]  - ./app/containers/HomePage/index.js
[HMR] App is up to date.
```

Nice!

----

I'll CC @jongold here, he was the one who initially reported the bug about the WebpackDevServer – would you mind taking a look at this Jon, does it fix your problem?

File: webpack/webpack/webpack.dev.babel.js [hunk 1]
```diff
@@ -36,6 +33,7 @@ module.exports = require('./webpack.base.babel')({
   // Add hot reloading
   plugins: [
     new webpack.HotModuleReplacementPlugin(), // Tell webpack we want hot reloading
+    new webpack.NoErrorsPlugin(),
```
Comment (mxstbr) [inline webpack/webpack/webpack.dev.babel.js]: What does the `NoErrorsPlugin` do and why do we need it? Don't we want to see errors?

**Comment by mxstbr** [comment]
After trying it out, this does not seem to fix the original problem of #113:

screen shot 2016-02-25 at 09 37 24

Any ideas @amorino?

File: webpack/webpack/webpack.dev.babel.js (hunk 1 above)
Comment (amorino) [inline webpack/webpack/webpack.dev.babel.js]: Because of [this](https://github.com/gaearon/react-hot-boilerplate/issues/1):

**Comment by amorino** [comment]
Mhm, with the historyApiFallback it should't happen. I think is because of the dots (1.2.3), with plain routes it also throws an error?

**Comment by mxstbr** [comment]
It works with plain routes, but the original point of switching to express was to fix the issues with the dots. Do you think you can get that to work?

**Comment by amorino** [comment]
Yep, I will search for that. My bet is on the historyapifallback. I will check the library.

**Comment by mxstbr** [comment]
Yep, that's it! See here: https://github.com/expressjs/express/issues/669#issuecomment-188949249

**Comment by mxstbr** [comment]
According to [#5 in connect-history-api-fallback](https://github.com/bripkens/connect-history-api-fallback/issues/5) this should be possible with the [`rewrite` option](https://github.com/bripkens/connect-history-api-fallback#rewrites), which was specifically introduced to fix this issue.

**Comment by mxstbr** [comment]
Any update on this @amorino? I'd love to finally merge this!

File: webpack/webpack/server.prod.js [hunk 2]
```diff
@@ -16,16 +24,25 @@ console.log(
 );
 console.log('Starting server from build folder...');
 
-new WebpackDevServer(webpack(config), { // Start a server
+app.use(historyApi({
+  verbose: false
+}));
+
+app.use(webpackDevMiddleware(compiler, {
```
Comment (grabbou) [inline webpack/webpack/server.prod.js]: Do we need webpack in production at all?

**Comment by grabbou** [comment]
Great work, going to try that out later today.

Few concerns/questions:
- Do we need Webpack Dev Server in production at all? It feels like we should enable gzip in there and just serve static files that were build upfront
- Can we just make one file, called "server" and based on the NODE_ENV, just add in appropriate middlewares? E.g. if development, add webpack-dev-middleware etc., otherwise, turn on production performance improvements

**Comment by mxstbr** [comment]
It appears there's a misconception here that comes from my bad naming practices - `server.prod.js` should actually not be used in production, ever. It solely exist to make `npm run serve` work and tunnels your local connection with `ngrok` to let you access your local application anywhere on the world – this should not be run in production at all, like, ever!

Thanks for the valuable feedback @grabbou, let's do the following:

- Add a **real, production ready `server.prod.js`**, enabling all performance improvements we can (e.g. `gzip`)
- **Rename `server.prod.js`** to something else (ideas? `server.serve.js` looks weird...) and have it start the production server and then tunnel *that* with `ngrok`.

**Comment by grabbou** [comment]
Yeah, I started thinking I might be mis-using that file to be fair! Maybe let's name it serve.dev.js and serve.prod.js? 

Anyway - I was thinking about adding top-level folder called `server` since in the future, once we move towards adding e.g. server-side rendering, that might look better. In my particular use-case, I have to handle form submission with POST request, so it feels better to live at the root instead of `internals` folder.

**Comment by mxstbr** [comment]
That sounds like a good idea, I like it!

**Comment by mxstbr** [comment]
I prefer @grabbou's #211, as it fits better with this boilerplate. Thanks so much for your work and input though @amorino, if you have any suggestions I'd love to hear them in #211!

**Comment by ASH-khan** [comment]
i am reading about the same issue from last 2 hours and all the solution posted here none is working :  here what i did in server.js file....
const express = require('express');
const path = require('path');
const port = 9990;
const app = express();

app.use(express.static(__dirname));

app.get('*',(req, res) => {
    res.sendFile(path.resolve(__dirname, 'index.html'));
});

app.listen(port);
console.log(`Server is Runing on Port # ${port}`);

//////////
app.get('*',(req, res) => {
    res.sendFile(path.resolve(__dirname, 'index.html'));
});
/////////

just add this function after the app.use() and working beautiful.
:)

**Pull request closed without merging**